### PR TITLE
ddl: fix dist ddl reorg table range endKey

### LINKED ISSUE #45652: Failed to create index with distribute framework after import data via lightning
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```
 gmhdbjd@gmhdbjd  ~/code/tidb/bin   master  ./tidb-lightning -c cfg.toml
Verbose debug logs will be written to tidb-lightning.log

tidb lightning exit successfully
 gmhdbjd@gmhdbjd  ~/code/tidb/bin   master  mysql -uroot -h127.0.0.1 -P4000
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 17
Server version: 5.7.25-TiDB-v7.3.0-alpha TiDB Server (Apache License 2.0) Community Edition, MySQL 8.0 compatible

Copyright (c) 2000, 2023, Oracle and/or its affiliates.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> set global tidb_enable_dist_task=1;
Query OK, 0 rows affected (0.03 sec)

mysql> use db;
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Database changed
mysql> select * from tb;
+------+------+
| id   | msg  |
+------+------+
|    1 | 1    |
+------+------+
1 row in set (0.00 sec)

mysql> show config where type = 'tikv' && name = 'storage.engine';
+------+-----------------+----------------+---------+
| Type | Instance        | Name           | Value   |
+------+-----------------+----------------+---------+
| tikv | 127.0.0.1:20160 | storage.engine | raft-kv |
+------+-----------------+----------------+---------+
1 row in set (0.01 sec)

mysql> create index idx on tb(id);



```

### 2. What did you expect to see? (Required)
create index succeed
### 3. What did you see instead (Required)
```

[2023/07/28 15:33:51.366 +08:00] [WARN] [dispatcher.go:379] ["gen dist-plan failed"] [error="PD returned no region, startKey: \"74800000000000006C5F728000000000000001\", endKey: \"74800000000000006C5F728000000000000001\", limit: 128"]
```
### 4. What is your TiDB version? (Required)
master

### PR DESCRIPTION
**Pull request by jiyfhust**
### What problem does this PR solve?

Issue Number: close #45652
Problem Summary:

### What is changed and how it works?

As the issue describe:

function getTableRange:
https://github.com/pingcap/tidb/blob/cf7ae4c3b917f81e43cbb2beab44abf841af52cc/ddl/reorg.go#L669C2-L669C2
return [startHandleKey, endHandleKey];
the reorg will use the range above to backfill index data.

Then splitTableRanges will split  [startHandleKey, endHandleKey] into small ranges.
https://github.com/pingcap/tidb/blob/cf7ae4c3b917f81e43cbb2beab44abf841af52cc/ddl/scheduler.go#L267C8-L267C8
the small ranges is like [a, b], [b.next, c],[c.next, d].

here use a function getRangeEndKey:
https://github.com/pingcap/tidb/blob/cf7ae4c3b917f81e43cbb2beab44abf841af52cc/ddl/backfilling.go#L562C17-L562C17
change small ranges to [a, b.next), [b.next, c.next), [c.next, d.next).
and it use a "endInclude" to mark case when it is [c.next d].

This pr:
1. fix the getTableRange, return [startHandleKey, endHandleKey.next)
2. split rangs into [a, b.next), [b.next, c.next), [c.next, d.next)
3. don't use getRangeEndKey to change the endKey
4. remove "endInclude"

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by zimulala — COMMENTED** [review]
Could you briefly describe the fix in the PR description?

**Comment by jiyfhust** [comment]
> Could you briefly describe the fix in the PR description?
/cc  @zimulala

**Comment by jiyfhust** [comment]
> Could you briefly describe the fix in the PR description?

/cc @zimulala

File: ddl/backfilling.go [hunk 1]
```diff
@@ -550,24 +538,12 @@ func getBatchTasks(t table.Table, reorgInfo *reorgInfo, kvRanges []kv.KeyRange,
...
-	job := reorgInfo.Job
 	//nolint:forcetypeassert
 	phyTbl := t.(table.PhysicalTable)
-	jobCtx := reorgInfo.d.jobContext(job.ID)
-	for i, keyRange := range kvRanges {
+	for _, keyRange := range kvRanges {
 		taskID := taskIDAlloc.alloc()
 		startKey := keyRange.StartKey
 		endKey := keyRange.EndKey
-		endK, err := getRangeEndKey(jobCtx, reorgInfo.d.store, job.Priority, prefix, keyRange.StartKey, endKey)
```
Comment (tangenta) [inline ddl/backfilling.go:562]: This seems to lead to performance regression mentioned in https://github.com/pingcap/tidb/issues/26683. @zimulala do we have relevant integration test about this issue?

Comment (jiyfhust) [reply ddl/backfilling.go:562]: Let me take a look at this problem.

**Comment by jiyfhust** [comment]
it seem to be regionCache bug................

**Comment by jiyfhust** [comment]
the LoadRegionsInKeyRange return range [start_key,end_key], not [start_key, end_key)
https://github.com/tikv/client-go/blob/cbfd64018c2194af29694547bfcad1972cb24512/internal/locate/region_cache.go#L994C3-L994C3

it calls BatchLoadRegionsWithKeyRange
https://github.com/tikv/client-go/blob/cbfd64018c2194af29694547bfcad1972cb24512/internal/locate/region_cache.go#L1018

but scanRegions return range [start_key, end_key)
https://github.com/tikv/client-go/blob/cbfd64018c2194af29694547bfcad1972cb24512/internal/locate/region_cache.go#L1336

/cc @tangenta @zimulala

**Comment by tangenta** [comment]
> it seem to be regionCache bug................

I didn't find the failed test. Could you provide the test name?

**Comment by jiyfhust** [comment]
> 

Sorry, just ignore it.

The code annotation mislead me.
https://github.com/tikv/client-go/blob/cbfd64018c2194af29694547bfcad1972cb24512/internal/locate/region_cache.go#L993
```
// LoadRegionsInKeyRange lists regions in [start_key,end_key].
func (c *RegionCache) LoadRegionsInKeyRange(bo *retry.Backoffer, startKey, endKey []byte) (regions []*Region, err error) {
	var batchRegions []*Region
	for {
		batchRegions, err = c.Bat
....
....
```
it is not [start_key,end_key], but actually [start_key,end_key)，so it is not a "tikv/client-go" bug which  i said regionCache above.

So there is no problems now.

File: ddl/backfilling.go [hunk 2]
```diff
@@ -555,11 +543,17 @@ func getBatchTasks(t table.Table, reorgInfo *reorgInfo, kvRanges []kv.KeyRange,
...
+	for _, keyRange := range kvRanges {
 		taskID := taskIDAlloc.alloc()
 		startKey := keyRange.StartKey
+		if len(startKey) == 0 {
+			startKey = prefix
+		}
 		endKey := keyRange.EndKey
-		endK, err := getRangeEndKey(jobCtx, reorgInfo.d.store, job.Priority, prefix, keyRange.StartKey, endKey)
+		if len(endKey) == 0 {
+			endKey = prefix.PrefixNext()
```
Comment (tangenta) [inline ddl/backfilling.go:554]: I just found that index ID is not included :(

`t.IndexPrefix()` is the key like `t[tableID]_i`. If we use `t.IndexPrefix().PrefixNext()` here, `endKey` will be `t[tableID]_j`?

As a result, we will scan the index record from **all** indexes.

**Comment by tangenta** [comment]
> I just found that index ID is not included :(
> 
> `t.IndexPrefix()` is the key like `t[tableID]_i`. If we use `t.IndexPrefix().PrefixNext()` here, `endKey` will be `t[tableID]_j`?
> 
> As a result, we will scan the index records from **all** indexes.

@jiyfhust Let me fix this issue in another PR, which will include some tests. Could you revert [47f36bd](https://github.com/pingcap/tidb/pull/45679/commits/47f36bd0831e6affa76ae1c0942752627f9600f4)? Rest LGTM.

**Comment by jiyfhust** [comment]
> 

done

**Comment by jiyfhust** [comment]
> @jiyfhust Let me fix this issue in another PR, which will include some tests. Could you revert [47f36bd](https://github.com/pingcap/tidb/pull/45679/commits/47f36bd0831e6affa76ae1c0942752627f9600f4)? Rest LGTM.

@tangenta done.

**Review by tangenta — APPROVED** [review]

**Review by zimulala — APPROVED** [review]
LGTM

**Pull request merged**